File: app/src/main/java/com/lnstretch/MainActivity.java
package com.lnstretch;

import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.widget.SeekBar;
import android.widget.Toast;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;
import com.lnstretch.databinding.ActivityMainBinding;
import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import rikka.shizuku.Shizuku;

public class MainActivity extends AppCompatActivity {

    private static final String PREFS      = "LNStretchPrefs";
    private static final String KEY_PCT    = "stretch_percent";
    private static final String FF_PKG     = "com.dts.freefireth";
    private static final String FF_MAX     = "com.dts.freefiremax";
    private static final int    SHIZUKU_REQ = 1001;

    private ActivityMainBinding b;
    private SharedPreferences   prefs;
    private int     pct       = 20;
    private boolean stretched = false;
    private int origW, origH, origDpi;

    @Override
    protected void onCreate(Bundle s) {
        super.onCreate(s);
        b = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(b.getRoot());
        prefs  = getSharedPreferences(PREFS, MODE_PRIVATE);
        pct    = prefs.getInt(KEY_PCT, 20);
        getMetrics();
        setupUI();
        checkShizuku();
    }

    // ── pega métricas reais da tela ──────────────────────────────
    private void getMetrics() {
        android.util.DisplayMetrics dm = getResources().getDisplayMetrics();
        origW   = dm.widthPixels;
        origH   = dm.heightPixels;
        origDpi = (int) dm.densityDpi;
    }

    // ── configura todos os controles da UI ───────────────────────
    private void setupUI() {
        b.seekBarStretch.setMax(30);
        b.seekBarStretch.setProgress(pct - 10);
        b.tvStretchValue.setText(pct + "%");

        b.seekBarStretch.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            public void onProgressChanged(SeekBar sk, int p, boolean f) {
                pct = p + 10;
                b.tvStretchValue.setText(pct + "%");
                prefs.edit().putInt(KEY_PCT, pct).apply();
            }
            public void onStartTrackingTouch(SeekBar sk) {}
            public void onStopTrackingTouch(SeekBar sk) {}
        });

        b.btnStretch.setOnClickListener(v   -> { if (checkOk()) stretch(); });
        b.btnUnstretch.setOnClickListener(v -> { if (checkOk()) unstretch(); });
        b.btnOptimize.setOnClickListener(v  -> { if (checkOk()) optimize(); });
        b.btnOpenFF.setOnClickListener(v    -> openFF());
        updateStatus();
    }

    // ── verifica estado do Shizuku ───────────────────────────────
    private void checkShizuku() {
        boolean ok = Shizuku.pingBinder();
        b.tvShizukuStatus.setText(ok ? "✅ Shizuku: Ativo" : "❌ Shizuku: Inativo — Abra o app Shizuku");
        b.tvShizukuStatus.setTextColor(ContextCompat.getColor(this,
                ok ? R.color.green_active : R.color.red_inactive));
    }

    private boolean checkOk() {
        if (!Shizuku.pingBinder()) {
            toast("⚠️ Shizuku não está ativo! Abra o app Shizuku.");
            return false;
        }
        if (Shizuku.checkSelfPermission() != PackageManager.PERMISSION_GRANTED) {
            Shizuku.requestPermission(SHIZUKU_REQ);
            return false;
        }
        return true;
    }

    // ── ESTICAR ──────────────────────────────────────────────────
    private void stretch() {
        new Thread(() -> {
            try {
                float factor = 1f + pct / 100f;
                int newW = (int)(origW * factor);
                int newDpi = Math.max((int)(origDpi / factor), 280);

                // Usa Shizuku.newProcess() para rodar como shell privilegiado
                runShizuku("wm size " + newW + "x" + origH);
                runShizuku("wm density " + newDpi);

                stretched = true;
                runOnUiThread(() -> { updateStatus(); toast("✅ Tela esticada " + pct + "%!"); });
            } catch (Exception e) {
                runOnUiThread(() -> toast("Erro ao esticar: " + e.getMessage()));
            }
        }).start();
    }

    // ── DESISTICAR ───────────────────────────────────────────────
    private void unstretch() {
        new Thread(() -> {
            try {
                runShizuku("wm size reset");
                runShizuku("wm density reset");

                stretched = false;
                runOnUiThread(() -> { updateStatus(); toast("✅ Tela restaurada!"); });
            } catch (Exception e) {
                runOnUiThread(() -> toast("Erro ao restaurar: " + e.getMessage()));
            }
        }).start();
    }

    // ── OTIMIZAR FF ──────────────────────────────────────────────
    private void optimize() {
        new Thread(() -> {
            try {
                String pkg = ffPkg();
                if (pkg == null) { runOnUiThread(() -> toast("❌ Free Fire não encontrado.")); return; }

                runShizuku("pm clear " + pkg);
                runShizuku("settings put global window_animation_scale 0.5");
                runShizuku("settings put global transition_animation_scale 0.5");
                runShizuku("settings put global animator_duration_scale 0.5");

                runOnUiThread(() -> toast("✅ Free Fire otimizado!"));
            } catch (Exception e) {
                runOnUiThread(() -> toast("Erro: " + e.getMessage()));
            }
        }).start();
    }

    // ── ABRIR FF ─────────────────────────────────────────────────
    private void openFF() {
        String pkg = ffPkg();
        if (pkg == null) { toast("❌ Free Fire não instalado."); return; }
        Intent i = getPackageManager().getLaunchIntentForPackage(pkg);
        if (i != null) { i.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK); startActivity(i); }
        else toast("❌ Não foi possível abrir o Free Fire.");
    }

    // ── EXECUTOR VIA SHIZUKU ─────────────────────────────────────
    // Usa Shizuku.newProcess() — roda cada comando como root/shell
    private void runShizuku(String command) throws Exception {
        // split em array para evitar problemas com espaços
        String[] cmdArray = new String[]{"sh", "-c", command};
        Process process = Shizuku.newProcess(cmdArray, null, null);
        int exit = process.waitFor();

        // lê stderr para debug (ignorado silenciosamente em produção)
        InputStream err = process.getErrorStream();
        BufferedReader br = new BufferedReader(new InputStreamReader(err));
        StringBuilder sb = new StringBuilder();
        String line;
        while ((line = br.readLine()) != null) sb.append(line);
        process.destroy();

        if (exit != 0 && sb.length() > 0) {
            throw new Exception("cmd falhou [" + exit + "]: " + sb.toString());
        }
    }

    // ── HELPERS ──────────────────────────────────────────────────
    private String ffPkg() {
        PackageManager pm = getPackageManager();
        try { pm.getPackageInfo(FF_PKG, 0); return FF_PKG; } catch (Exception ignored) {}
        try { pm.getPackageInfo(FF_MAX, 0); return FF_MAX; } catch (Exception ignored) {}
        return null;
    }

    private void updateStatus() {
        if (stretched) {
            b.tvStretchStatus.setText("🎯 Tela ESTICADA (" + pct + "%)");
            b.tvStretchStatus.setTextColor(ContextCompat.getColor(this, R.color.orange_stretched));
            b.btnStretch.setEnabled(false);
            b.btnUnstretch.setEnabled(true);
        } else {
            b.tvStretchStatus.setText("📱 Tela Normal");
            b.tvStretchStatus.setTextColor(ContextCompat.getColor(this, R.color.green_active));
            b.btnStretch.setEnabled(true);
            b.btnUnstretch.setEnabled(false);
        }
    }

    private void toast(String msg) {
        new Handler(Looper.getMainLooper()).post(() ->
                Toast.makeText(this, msg, Toast.LENGTH_SHORT).show());
    }

    @Override
    protected void onResume() {
        super.onResume();
        checkShizuku();
    }
}

File: app/src/main/java/com/lnstretch/MainActivity.kt
package com.lnstretch

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.util.DisplayMetrics
import android.view.WindowManager
import android.widget.SeekBar
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.lnstretch.databinding.ActivityMainBinding
import rikka.shizuku.Shizuku
import rikka.shizuku.ShizukuRemoteProcess

class MainActivity : AppCompatActivity() {

    private lateinit var binding: ActivityMainBinding

    private val ffPackages = listOf(
        "com.dts.freefireth",
        "com.dts.freefiremax"
    )

    private var selectedPercent: Int = 15

    private val shizukuPermissionListener = Shizuku.OnRequestPermissionResultListener { _, grantResult ->
        if (grantResult == PackageManager.PERMISSION_GRANTED) {
            showToast("✅ Shizuku autorizado!")
            updateShizukuStatus()
        } else {
            showToast("❌ Permissão negada pelo Shizuku")
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)
        Shizuku.addRequestPermissionResultListener(shizukuPermissionListener)
        setupSlider()
        setupButtons()
        updateStatus()
    }

    override fun onDestroy() {
        super.onDestroy()
        Shizuku.removeRequestPermissionResultListener(shizukuPermissionListener)
    }

    override fun onResume() {
        super.onResume()
        updateShizukuStatus()
    }

    private fun setupSlider() {
        // 0 = 10%, step 1% cada, max 30 = 40%
        binding.seekbarStretch.max = 30
        binding.seekbarStretch.progress = 5 // 15%
        updateSliderLabel(15)

        binding.seekbarStretch.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar?, progress: Int, fromUser: Boolean) {
                selectedPercent = 10 + progress
                updateSliderLabel(selectedPercent)
            }
            override fun onStartTrackingTouch(seekBar: SeekBar?) {}
            override fun onStopTrackingTouch(seekBar: SeekBar?) {}
        })
    }

    @SuppressLint("SetTextI18n")
    private fun updateSliderLabel(percent: Int) {
        val metrics = getRealMetrics()
        val newWidth = (metrics.widthPixels * (1 + percent / 100.0)).toInt()
        binding.tvSliderLabel.text = "+${percent}%   ${metrics.widthPixels}px → ${newWidth}px"
    }

    private fun setupButtons() {
        binding.btnStretch.setOnClickListener { applyStretch() }
        binding.btnUnstretch.setOnClickListener { removeStretch() }
        binding.btnOptimize.setOnClickListener { optimizeFF() }
        binding.btnOpenFF.setOnClickListener { openFreefire() }
        binding.btnShizuku.setOnClickListener { requestShizukuPermission() }
    }

    @SuppressLint("SetTextI18n")
    private fun applyStretch() {
        if (!isShizukuReady()) {
            binding.tvStatus.text = "⚠️ Shizuku não autorizado!\nClique em AUTORIZAR SHIZUKU primeiro."
            showToast("Autorize o Shizuku primeiro!")
            return
        }
        val metrics = getRealMetrics()
        val origWidth = metrics.widthPixels
        val origHeight = metrics.heightPixels
        val newWidth = (origWidth * (1 + selectedPercent / 100.0)).toInt()
        val success = runShizukuCommand("wm size ${newWidth}x${origHeight}")
        if (success) {
            saveBackup(origWidth, origHeight, metrics.densityDpi)
            binding.tvStatus.text = "✅ Tela esticada +${selectedPercent}%!\n${origWidth}x${origHeight}  →  ${newWidth}x${origHeight}\n\n🎮 Abra o Free Fire agora!"
            showToast("Esticado +${selectedPercent}%! 🔥")
        } else {
            binding.tvStatus.text = "❌ Erro ao esticar. Verifique o Shizuku."
        }
    }

    @SuppressLint("SetTextI18n")
    private fun removeStretch() {
        if (!isShizukuReady()) {
            showToast("Autorize o Shizuku primeiro!")
            return
        }
        val prefs = getSharedPreferences("ffstretch_prefs", Context.MODE_PRIVATE)
        val origWidth = prefs.getInt("orig_width", 0)
        val origHeight = prefs.getInt("orig_height", 0)
        val cmd = if (origWidth > 0) "wm size ${origWidth}x${origHeight}" else "wm size reset"
        val success = runShizukuCommand(cmd)
        if (success) {
            clearBackup()
            binding.tvStatus.text = "✅ Tela restaurada!\n${origWidth}x${origHeight}"
            showToast("Tela restaurada!")
        } else {
            binding.tvStatus.text = "❌ Erro ao restaurar.\nTente: wm size reset via Shizuku."
        }
    }

    @SuppressLint("SetTextI18n")
    private fun optimizeFF() {
        binding.tvStatus.text = "⚡ Otimizando..."
        binding.btnOptimize.isEnabled = false
        Thread {
            var ffFound = false
            if (isShizukuReady()) {
                runShizukuCommand("am kill-all")
                ffPackages.forEach { pkg ->
                    try {
                        packageManager.getPackageInfo(pkg, 0)
                        ffFound = true
                        runShizukuCommand("am send-trim-memory $pkg COMPLETE")
                    } catch (_: PackageManager.NameNotFoundException) {}
                }
            }
            System.gc()
            Runtime.getRuntime().gc()
            val free = Runtime.getRuntime().freeMemory() / 1024 / 1024
            val total = Runtime.getRuntime().totalMemory() / 1024 / 1024
            runOnUiThread {
                binding.btnOptimize.isEnabled = true
                val ffStatus = if (ffFound) "✅ Free Fire otimizado" else "ℹ️ FF não encontrado"
                binding.tvStatus.text = "⚡ Pronto!\n$ffStatus\n• RAM livre: ${free}MB / ${total}MB\n• Apps background encerrados"
                showToast("FF otimizado! 🔥")
            }
        }.start()
    }

    @SuppressLint("SetTextI18n")
    private fun openFreefire() {
        var launched = false
        for (pkg in ffPackages) {
            try {
                val intent = packageManager.getLaunchIntentForPackage(pkg)
                if (intent != null) {
                    intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                    startActivity(intent)
                    launched = true
                    binding.tvStatus.text = "🎮 Abrindo Free Fire...\n✅ Tela +${selectedPercent}% ativa"
                    break
                }
            } catch (_: Exception) {}
        }
        if (!launched) {
            binding.tvStatus.text = "❌ Free Fire não encontrado!\nInstale o jogo primeiro."
            showToast("Free Fire não instalado!")
        }
    }

    private fun isShizukuReady(): Boolean {
        return try {
            Shizuku.pingBinder() && Shizuku.checkSelfPermission() == PackageManager.PERMISSION_GRANTED
        } catch (_: Exception) { false }
    }

    private fun requestShizukuPermission() {
        try {
            if (!Shizuku.pingBinder()) {
                showToast("Shizuku não está rodando!\nAbra o app Shizuku primeiro.")
                binding.tvStatus.text = "❌ Shizuku não encontrado!\nInstale e inicie o app Shizuku."
                return
            }
            if (Shizuku.checkSelfPermission() == PackageManager.PERMISSION_GRANTED) {
                showToast("Shizuku já está autorizado ✅")
                updateShizukuStatus()
                return
            }
            Shizuku.requestPermission(1001)
        } catch (e: Exception) {
            binding.tvStatus.text = "❌ Erro: ${e.message}\nInstale o app Shizuku."
        }
    }

    private fun runShizukuCommand(command: String): Boolean {
        return try {
            val process: ShizukuRemoteProcess = Shizuku.newProcess(
                arrayOf("sh", "-c", command), null, null
            )
            process.waitFor()
            true
        } catch (_: Exception) { false }
    }

    @SuppressLint("SetTextI18n")
    private fun updateStatus() {
        val metrics = getRealMetrics()
        val ffInstalled = ffPackages.any { pkg ->
            try { packageManager.getPackageInfo(pkg, 0); true }
            catch (_: PackageManager.NameNotFoundException) { false }
        }
        val ffStatus = if (ffInstalled) "✅ Free Fire detectado" else "❌ Free Fire não instalado"
        binding.tvStatus.text = "📱 Resolução: ${metrics.widthPixels}x${metrics.heightPixels}\n$ffStatus"
        updateShizukuStatus()
    }

    @SuppressLint("SetTextI18n")
    private fun updateShizukuStatus() {
        try {
            val running = Shizuku.pingBinder()
            val granted = running && Shizuku.checkSelfPermission() == PackageManager.PERMISSION_GRANTED
            binding.tvShizukuStatus.text = when {
                granted -> "🟢 Shizuku: Autorizado"
                running -> "🟡 Shizuku: Ativo (clique Autorizar)"
                else    -> "🔴 Shizuku: Não encontrado"
            }
            binding.btnShizuku.isEnabled = running && !granted
        } catch (_: Exception) {
            binding.tvShizukuStatus.text = "🔴 Shizuku: Não instalado"
        }
    }

    private fun getRealMetrics(): DisplayMetrics {
        val wm = getSystemService(Context.WINDOW_SERVICE) as WindowManager
        val metrics = DisplayMetrics()
        @Suppress("DEPRECATION")
        wm.defaultDisplay.getRealMetrics(metrics)
        return metrics
    }

    private fun saveBackup(width: Int, height: Int, dpi: Int) {
        getSharedPreferences("ffstretch_prefs", Context.MODE_PRIVATE).edit()
            .putInt("orig_width", width)
            .putInt("orig_height", height)
            .putInt("orig_dpi", dpi)
            .apply()
    }

    private fun clearBackup() {
        getSharedPreferences("ffstretch_prefs", Context.MODE_PRIVATE).edit().clear().apply()
    }

    private fun showToast(msg: String) = Toast.makeText(this, msg, Toast.LENGTH_SHORT).show()
}
